File: Engine/src/main/java/org/etieskrill/engine/scene/component/Container.java
package org.etieskrill.engine.scene.component;

import org.etieskrill.engine.graphics.Batch;
import org.etieskrill.engine.input.Key;
import org.etieskrill.engine.input.Keys;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.joml.Vector3f;

import static java.util.Objects.requireNonNull;
import static org.etieskrill.engine.scene.component.LayoutUtils.getPreferredNodePosition;

/**
 * A node with a single child.
 */
public class Container extends Node<Container> {

    protected @Nullable Node<?> child;

    public Container(@Nullable Node<?> child) {
        if (child != null) setChild(child);
    }

    @Override
    public void update(double delta) {
        if (child != null) child.update(delta);
    }

    @Override
    public void format() {
        if (!shouldFormat() || child == null) return;

        child.format();
        child.setPosition(getPreferredNodePosition(getSize(), child));
    }

    @Override
    public void render(@NotNull Batch batch) {
        if (getRenderedColour().w != 0) batch.renderBox(
                new Vector3f(getPosition(), 0),
                new Vector3f(getSize(), 0),
                getRenderedColour()
        );
        if (child != null) child.render(batch);
    }

    public void setChild(@NotNull Node<?> child) {
        invalidate();
        requireNonNull(child).setParent(this);
        this.child = child;
    }

    @Override
    public boolean handleHit(@NotNull Key button, Keys.@NotNull Action action, double posX, double posY) {
        if (!doesHit(posX, posY)) return false;
        if (child == null) return false;
        return child.handleHit(button, action, posX, posY); //container itself is not hittable
    }

    @Override
    public boolean handleDrag(double deltaX, double deltaY, double posX, double posY) {
        if (!doesHit(posX, posY)) return false;
        if (child == null) return false;
        return child.handleDrag(deltaX, deltaY, posX, posY);
    }

}

File: Engine/src/main/java/org/etieskrill/engine/scene/component/container/Container.kt
package org.etieskrill.engine.scene.component.container

import org.etieskrill.engine.graphics.Batch
import org.etieskrill.engine.input.Key
import org.etieskrill.engine.input.Keys
import org.etieskrill.engine.scene.component.Node
import org.etieskrill.engine.scene.component.getPreferredNodePosition
import org.joml.Vector3f

/**
 * A node with a single child.
 */
open class Container(child: Node<*>?) : Node<Container>() {

    var child: Node<*>? = child
        set(value) {
            invalidate()
            child?.parent = this
            field = value
        }

    init {
        this.child = child //setter is not called on property initialisation
    }

    override fun update(delta: Double) {
        child?.update(delta)
    }

    override fun format() {
        if (!shouldFormat()) return

        child?.run {
            format()
            position = getPreferredNodePosition(size, this)
        }
    }

    override fun render(batch: Batch) {
        if (renderedColour.w != 0f) {
            batch.renderBox(
                Vector3f(position, 0f),
                Vector3f(size, 0f),
                renderedColour
            )
        }

        child?.render(batch)
    }

    override fun handleHit(button: Key, action: Keys.Action, posX: Double, posY: Double) = when {
        !doesHit(posX, posY) -> false
        child == null -> false
        else -> child!!.handleHit(button, action, posX, posY)
    }

    override fun handleKey(key: Key, action: Keys.Action) =
        child?.handleKey(key, action) ?: false

    override fun handleHover(posX: Double, posY: Double) = when {
        !doesHit(posX, posY) -> false
        child == null -> false
        else -> child!!.handleHover(posX, posY)
    }

    override fun handleDrag(deltaX: Double, deltaY: Double, posX: Double, posY: Double) = when {
        !doesHit(posX, posY) -> false
        child == null -> false
        else -> child!!.handleDrag(deltaX, deltaY, posX, posY)
    }

}